libethereum/ValidationSchemes.cpp

```diff
@@ -0,0 +1,141 @@
+string const c_minimumDifficulty = "minimumDifficulty";
+string const c_durationLimit = "durationLimit";
+string const c_chainID = "chainID";
+string const c_networkID = "networkID";
+string const c_allowFutureBlocks = "allowFutureBlocks";
+
+void validateConfigJson(js::mObject const& _obj)
+{
+    requireJsonFields(_obj, "ChainParams::loadConfig",
+        {{c_sealEngine, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_params, {{js::obj_type}, JsonFieldPresence::Required}},
+            {c_genesis, {{js::obj_type}, JsonFieldPresence::Required}},
+            {c_accounts, {{js::obj_type}, JsonFieldPresence::Required}}});
+
+    requireJsonFields(_obj.at(c_genesis).get_obj(), "ChainParams::loadConfig::genesis",
+        {{c_author, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_nonce, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_gasLimit, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_timestamp, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_difficulty, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_extraData, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_mixHash, {{js::str_type}, JsonFieldPresence::Required}},
+            {c_parentHash, {{js::str_type}, JsonFieldPresence::Optional}}});
+
+    js::mObject const& accounts = _obj.at(c_accounts).get_obj();
+    for (auto const& acc : accounts)
+        validateAccountObj(acc.second.get_obj());
+}
+
+void validateAccountMapObj(js::mObject const& _obj)
+{
+    // A map object with possibly defined fields
+    requireJsonFields(_obj, "validateAccountMapObj",
+        {{c_storage, {{js::obj_type}, JsonFieldPresence::Optional}},
+            {c_balance, {{js::str_type}, JsonFieldPresence::Optional}},
+            {c_nonce, {{js::str_type}, JsonFieldPresence::Optional}},
+            {c_code, {{js::str_type}, JsonFieldPresence::Optional}},
+            {c_precompiled, {{js::obj_type}, JsonFieldPresence::Optional}},
+            {c_shouldnotexist, {{js::str_type}, JsonFieldPresence::Optional}},
+            {c_wei, {{js::str_type}, JsonFieldPresence::Optional}}});
+}
+
+void validateAccountObj(js::mObject const& _obj)
+{
+    if (_obj.count(c_precompiled))
+    {
+        // A precompiled contract
+        requireJsonFields(_obj, "validateAccountObj",
+            {{c_precompiled, {{js::obj_type}, JsonFieldPresence::Required}},
+                {c_wei, {{js::str_type}, JsonFieldPresence::Optional}}});
+    }
+    else if (_obj.size() == 1)
+    {
+        // A genesis account with only balance set
+        if (_obj.count(c_balance))
+            requireJsonFields(_obj, "validateAccountObj",
+                {{c_balance, {{js::str_type}, JsonFieldPresence::Required}}});
+        else
+            requireJsonFields(_obj, "validateAccountObj",
+                {{c_wei, {{js::str_type}, JsonFieldPresence::Required}}});
+    }
+    else
+    {
+        // A standart account with all fields
```

Comment:
In which configs are they used?

Reply:
Ah it's allowed to have in genesis accounts either balance only or all 4 fields? ok then

Reply:
In tests also